src/CrsGraph.C

```diff
@@ -0,0 +1,1031 @@
+  kokkos_parallel_for("Nalu::CrsGraph::beginConstructionA", buckets.size(), [&] (const int& ib) {
+    stk::mesh::Bucket & b = *buckets[ib];
+    const stk::mesh::Bucket::size_type length = b.size();
+    //KOKKOS: intra BucketLoop parallel reduce
+    for ( stk::mesh::Bucket::size_type k = 0 ; k < length ; ++k ) {
+
+      // get node
+      stk::mesh::Entity node = b[k];
+      int status = getDofStatus(node);
+
+      if (status & DS_SkippedDOF)
+        continue;
+
+      if (status & DS_OwnedDOF) {
+        numNodes++;
+        numOwnedNodes++;
+      }
+
+      if (status & DS_SharedNotOwnedDOF) {
+        numNodes++;
+        numSharedNotOwnedNotLocallyOwned++;
+      }
+
+      if (status & DS_GhostedDOF) {
+        numGhostNodes++;
+      }
+    }
+  });
+
+  maxOwnedRowId_ = numOwnedNodes * numDof_;
+  maxSharedNotOwnedRowId_ = numNodes * numDof_;
+
+  // Next, grab all the global ids, owned first, then sharedNotOwned.
+
+  // Also, we'll build up our own local id map. Note: first we number
+  // the owned nodes then we number the sharedNotOwned nodes.
+
+  // make separate arrays that hold the owned and sharedNotOwned gids
+  std::vector<stk::mesh::Entity> owned_nodes, shared_not_owned_nodes;
+  owned_nodes.reserve(numOwnedNodes);
+  shared_not_owned_nodes.reserve(numSharedNotOwnedNotLocallyOwned);
+
+  std::vector<GlobalOrdinal> ownedGids, sharedNotOwnedGids;
+  ownedGids.reserve(maxOwnedRowId_);
+  sharedNotOwnedGids.reserve(numSharedNotOwnedNotLocallyOwned*numDof_);
+  sharedPids_.reserve(sharedNotOwnedGids.capacity());
+
+  // owned first:
+  for(const stk::mesh::Bucket* bptr : buckets) {
+    const stk::mesh::Bucket & b = *bptr;
+    for ( stk::mesh::Entity entity : b ) {
+      int status = getDofStatus(entity);
+      if (!(status & DS_SkippedDOF) && (status & DS_OwnedDOF))
+        owned_nodes.push_back(entity);
+    }
+  }
+
+  std::sort(owned_nodes.begin(), owned_nodes.end(), CompareEntityById(bulkData, realm_.naluGlobalId_) );
+
+  // use the Contiguous Map constructor. 
+
+   const Teuchos::RCP<LinSys::Comm> tpetraComm = Teuchos::rcp(new LinSys::Comm(bulkData.parallel()));
+  ownedRowsMap_ = Teuchos::rcp(new LinSys::Map(Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid(),
+                                               maxOwnedRowId_, // must have *numDof_ 
+                                               1,
+                                               tpetraComm));
+  myLIDs_.clear();
+  myLIDs_.reserve(maxOwnedRowId_ +  numSharedNotOwnedNotLocallyOwned*numDof_);
+
+  LocalOrdinal localId = 0;
+  int gstart = 0;
+
+  GlobalOrdinal gomin   = ownedRowsMap_->getMinGlobalIndex();
+
+  for(stk::mesh::Entity entity : owned_nodes) {
+    const stk::mesh::EntityId entityId = *stk::mesh::field_data(*realm_.naluGlobalId_, entity);
+    myLIDs_[entityId] = numDof_*localId++;
+    auto *  thisgid = stk::mesh::field_data(*realm_.tpetGlobalId_, entity);
+    auto basegid = gomin + numDof_ * gstart;
```

Comment:
The dependency on `numDof_` here means that the tpetra global id field can't be used in the `CrsGraph::finalizeGraph()` step unless the last call to a `CrsGraph::beginConstruction()` method used the same `numDof_` parameter.

This I think is the issue with the failing tests.  We could bypass this issue though for now by making sure that a `CrsGraph` only ever calls `CrsGraph::finalizeGraph()` once.